flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/MetadataTablePlanner.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.iceberg.flink.source.FlinkSplitPlanner;
+import org.apache.iceberg.flink.source.ScanContext;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplit;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplitSerializer;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.ThreadPools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Plans the splits to read a metadata table content. */
+@Internal
+public class MetadataTablePlanner extends ProcessFunction<Trigger, MetadataTablePlanner.SplitInfo> {
+  private static final Logger LOG = LoggerFactory.getLogger(MetadataTablePlanner.class);
+
+  private final String taskName;
+  private final int taskIndex;
+  private final TableLoader tableLoader;
+  private final int workerPoolSize;
+  private final ScanContext scanContext;
+  private transient ExecutorService workerPool;
+  private transient Counter errorCounter;
+  private transient Table table;
+  private transient IcebergSourceSplitSerializer splitSerializer;
+  private final MetadataTableType metadataTableType;
+
+  public MetadataTablePlanner(
+      String taskName,
+      int taskIndex,
+      TableLoader tableLoader,
+      ScanContext scanContext,
+      MetadataTableType metadataTableType,
+      int workerPoolSize) {
+    Preconditions.checkNotNull(taskName, "Task name should no be null");
+    Preconditions.checkNotNull(tableLoader, "Table should no be null");
+    Preconditions.checkArgument(scanContext.isStreaming(), "Streaming should be set to true");
+
+    this.taskName = taskName;
+    this.taskIndex = taskIndex;
+    this.tableLoader = tableLoader;
+    this.scanContext = scanContext;
+    this.workerPoolSize = workerPoolSize;
+    this.metadataTableType = metadataTableType;
+  }
+
+  @Override
+  public void open(OpenContext openContext) throws Exception {
+    tableLoader.open();
+    Table tableOri = tableLoader.loadTable();
+    this.table = MetadataTableUtils.createMetadataTableInstance(tableOri, metadataTableType);
+    this.workerPool =
+        ThreadPools.newFixedThreadPool(table.name() + "-table-planner", workerPoolSize);
+    this.splitSerializer = new IcebergSourceSplitSerializer(scanContext.caseSensitive());
+    this.errorCounter =
+        TableMaintenanceMetrics.groupFor(getRuntimeContext(), tableOri.name(), taskName, taskIndex)
+            .counter(TableMaintenanceMetrics.ERROR_COUNTER);
+  }
+
+  @Override
+  public void processElement(Trigger trigger, Context ctx, Collector<SplitInfo> out)
+      throws Exception {
+    try {
+      table.refresh();
+      for (IcebergSourceSplit split :
+          FlinkSplitPlanner.planIcebergSourceSplits(table, scanContext, workerPool)) {
+        out.collect(new SplitInfo(splitSerializer.getVersion(), splitSerializer.serialize(split)));
+      }
+    } catch (Exception e) {
+      LOG.error("Exception planning scan for {} at {}", table, ctx.timestamp(), e);
```

Comment:
This should be `warn`. The Job will not stop, but we would like to highlight this to the maintainer